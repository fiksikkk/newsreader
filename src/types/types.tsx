import { NativeStackScreenProps } from '@react-navigation/native-stack';

export type StackParamList = {
  Home: undefined;
  Character: { userid: number | string };
  CreateNewCharacter: { createCharacter: boolean};
};

export type HomeProps = NativeStackScreenProps<StackParamList, 'Home'>;
export type CharacterProps = NativeStackScreenProps<StackParamList,'Character'>;
export type CreacteCharacterProps = NativeStackScreenProps<StackParamList, 'CreateNewCharacter'>;

export interface Character {
  name: string;
  image: string;
  id: number;
}

interface Info {
  next: number;
  prev: number;
  pages: number;
  count: number;
}

export interface Characters {
  characters: { results: Character[]; info: Info };
}

export interface GetPostsVariables {
  page: number
}