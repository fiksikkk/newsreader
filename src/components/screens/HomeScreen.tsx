import DisplayPosts from './Home-displayPosts';
import {SafeAreaView, ScrollView, StyleSheet, View} from 'react-native';
import {HomeProps} from '../../types/types';

const HomeScreen = ({navigation}: HomeProps) => {
  return (
    <SafeAreaView>
      <ScrollView>
        <View style={styles.container}>
          <DisplayPosts navigation={navigation} />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignItems: 'center',
  },
});

export default HomeScreen;
