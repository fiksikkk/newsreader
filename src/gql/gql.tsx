import {gql} from '@apollo/client';

export const GET_POSTS = gql`
  query {
    characters {
      info {
        count
      }
      results {
        id
        name
        image
      }
    }
  }
`;

export const GET_CHARACTER = (id: number | string) => {
  return gql`
    query {
      character(id: ${id}) {
        name
        status
        species
        type
        gender
        image
      }      
    }
  `;
};
