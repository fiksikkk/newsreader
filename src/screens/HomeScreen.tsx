import React from 'react';
import { useState } from 'react';

import { useQuery } from '@apollo/client';
import { SafeAreaView, Text } from 'react-native';

import DisplayCards from '../components/displayCards';
import Header from '../components/header';
import { GET_POSTS } from '../gql/gql';
import { Characters, GetPostsVariables, HomeProps } from '../types/types';

const HomeScreen = ({ navigation }: HomeProps) => {
  const [page] = useState(1);

  const viewCharacter = (id: number) => {
    navigation.navigate('Character', { userid: id });
  };

  const { loading, error, data, fetchMore, refetch } = useQuery<Characters, GetPostsVariables>(GET_POSTS,
    { variables: { page }, notifyOnNetworkStatusChange: true });

  if (!data?.characters?.results?.length && loading) {
    return <Text testID="progress">Loading...</Text>;
  }
  if (error) {
    return <Text testID="error">Error : {error.message}</Text>;
  }
  if (!data?.characters?.results?.length) {
    return <Text testID="noData">Sorry, there is no Characters to show  </Text>
  }

  return (
    <SafeAreaView>
      <Header
        count={data.characters.info.count}
        results={data.characters.results}
      />

      <DisplayCards viewCharacter={viewCharacter} data={data} fetchMore={fetchMore} loading={loading} refetch={refetch} />

    </SafeAreaView>
  );
};

export default HomeScreen;
