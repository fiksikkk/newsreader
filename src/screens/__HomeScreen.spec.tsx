import React from 'react';

import { MockedProvider } from '@apollo/client/testing';
import { render, waitFor } from '@testing-library/react-native';

import { GET_POSTS } from '../gql/gql';
import HomeScreen from './HomeScreen';
import { HomeProps } from '../types/types';

describe('Should render HomeScreen', () => {
  const navigation = { navigate: jest.fn() } as unknown as HomeProps['navigation'];
  const route = { route: jest.fn() } as unknown as HomeProps['route'];

  it('should render and show progress on loading', () => {
    const utils = render(
      <MockedProvider addTypename={false} mocks={[mock(1)]}>
        <HomeScreen navigation={navigation} route={route} />
      </MockedProvider>,
    );

    expect(utils.queryByTestId('progress')).toBeTruthy();
  });

  it('should render 5 cards when loading is done', async () => {
    const utils = render(
      <MockedProvider addTypename={false} mocks={[mock(1)]}>
        <HomeScreen navigation={navigation} route={route} />
      </MockedProvider>,
    );

    await waitFor(() => [
      expect(utils.queryByTestId('progress')).toBeFalsy(),
    ]);

    expect(utils.queryAllByTestId('card').length).toBe(5);
    expect(utils.toJSON()).toMatchSnapshot();
  });
});

const mock = (page: number) => {
  return {
    request: {
      query: GET_POSTS, variables: { page },
    },
    result: {
      data: {
        characters: {
          info: {
            count: 826,
            next: 2,
            prev: null,
            pages: 42,
          },
          results: [
            {
              id: '1',
              name: 'Rick Sanchez',
              image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
            },
            {
              id: '2',
              name: 'Morty Smith',
              image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
            },
            {
              id: '3',
              name: 'Summer Smith',
              image: 'https://rickandmortyapi.com/api/character/avatar/3.jpeg',
            },
            {
              id: '4',
              name: 'Beth Smith',
              image: 'https://rickandmortyapi.com/api/character/avatar/4.jpeg',
            },
            {
              id: '5',
              name: 'Jerry Smith',
              image: 'https://rickandmortyapi.com/api/character/avatar/5.jpeg',
            }
          ],
        },
      },
    },
  };
};
