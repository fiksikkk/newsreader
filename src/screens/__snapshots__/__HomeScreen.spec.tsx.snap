// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render HomeScreen should render 5 cards when loading is done 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "alignItems": "center",
        "color": "black",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "black",
        }
      }
    >
      Here is 
      826
       characters
    </Text>
    <Text
      style={
        {
          "color": "black",
        }
      }
    >
      Displayed: 
      5
    </Text>
  </View>
  <RCTScrollView
    ListFooterComponent={null}
    contentContainerStyle={
      {
        "alignItems": "center",
        "flexDirection": "column",
        "paddingBottom": 200,
      }
    }
    data={
      [
        {
          "id": "1",
          "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
          "name": "Rick Sanchez",
        },
        {
          "id": "2",
          "image": "https://rickandmortyapi.com/api/character/avatar/2.jpeg",
          "name": "Morty Smith",
        },
        {
          "id": "3",
          "image": "https://rickandmortyapi.com/api/character/avatar/3.jpeg",
          "name": "Summer Smith",
        },
        {
          "id": "4",
          "image": "https://rickandmortyapi.com/api/character/avatar/4.jpeg",
          "name": "Beth Smith",
        },
        {
          "id": "5",
          "image": "https://rickandmortyapi.com/api/character/avatar/5.jpeg",
          "name": "Jerry Smith",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="container"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#841584",
              "borderRadius": 10,
              "borderWidth": 2,
              "flexDirection": "column",
              "height": 280,
              "margin": 10,
              "padding": 25,
            }
          }
          testID="card"
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Image
              src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
              style={
                {
                  "height": 200,
                  "width": 200,
                }
              }
            />
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "black",
                  "fontSize": 24,
                  "width": 200,
                }
              }
            >
              Rick Sanchez
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#841584",
              "borderRadius": 10,
              "borderWidth": 2,
              "flexDirection": "column",
              "height": 280,
              "margin": 10,
              "padding": 25,
            }
          }
          testID="card"
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Image
              src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
              style={
                {
                  "height": 200,
                  "width": 200,
                }
              }
            />
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "black",
                  "fontSize": 24,
                  "width": 200,
                }
              }
            >
              Morty Smith
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#841584",
              "borderRadius": 10,
              "borderWidth": 2,
              "flexDirection": "column",
              "height": 280,
              "margin": 10,
              "padding": 25,
            }
          }
          testID="card"
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Image
              src="https://rickandmortyapi.com/api/character/avatar/3.jpeg"
              style={
                {
                  "height": 200,
                  "width": 200,
                }
              }
            />
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "black",
                  "fontSize": 24,
                  "width": 200,
                }
              }
            >
              Summer Smith
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#841584",
              "borderRadius": 10,
              "borderWidth": 2,
              "flexDirection": "column",
              "height": 280,
              "margin": 10,
              "padding": 25,
            }
          }
          testID="card"
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Image
              src="https://rickandmortyapi.com/api/character/avatar/4.jpeg"
              style={
                {
                  "height": 200,
                  "width": 200,
                }
              }
            />
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "black",
                  "fontSize": 24,
                  "width": 200,
                }
              }
            >
              Beth Smith
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#841584",
              "borderRadius": 10,
              "borderWidth": 2,
              "flexDirection": "column",
              "height": 280,
              "margin": 10,
              "padding": 25,
            }
          }
          testID="card"
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Image
              src="https://rickandmortyapi.com/api/character/avatar/5.jpeg"
              style={
                {
                  "height": 200,
                  "width": 200,
                }
              }
            />
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "black",
                  "fontSize": 24,
                  "width": 200,
                }
              }
            >
              Jerry Smith
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
